src/Spectre.libStatistics/Math.h

```diff
@@ -380,4 +381,43 @@ std::vector<DataType> build(size_t size, Generator generator)
     return result;
 }
 
+/// <summary>
+/// Calculates differences between adjacent elements of given container.
+/// The function is based upon Matlab function diff():
+/// https://www.mathworks.com/help/matlab/ref/diff.html
+/// but it can handle only one dimensional data.
+/// </summary>
+/// <param name="data">The container.</param>
+/// <param name="order">Order of difference operation (how many times the difference is executed).</param>
+/// <returns>Vector of calculated differences, of size: length_of_data - order.</returns>
+template <class DataType>
+std::vector<typename std::remove_const<DataType>::type> differentiate_unsafe(gsl::span<DataType> data, uint16_t order = 1)
+{
+    std::vector<std::remove_const<DataType>::type> result(data.begin(), data.end());
+    for (auto i = 0u; i < order; ++i)
+    {
+        for (auto j = 0u; j < result.size() - 1; ++j)
+            result[j] = result[j + 1] - result[j];
+        result.resize(result.size() - 1);
+    }
+    return result;
+}
+
+/// <summary>
+/// Safe version of <see cref="differentiate_unsafe"/>.
+/// Validates the input before performing algorithm.
+/// </summary>
+/// <param name="data">The container.</param>
+/// <param name="order">Order of difference operation (how many times the difference is executed).</param>
+/// <returns>Vector of calculated differences, of size: length_of_data - order.</returns>
+template <class DataType>
+std::vector<typename std::remove_const<DataType>::type> differentiate(gsl::span<DataType> data, uint16_t order = 1)
+{
+    size_t length = data.length();
+    if (length < 2)
+        throw std::invalid_argument("Data must have length greater than two in order to perform differentiation.");
```

Comment:
We have our exception base to differentiate our exceptions of the unhandled ones. We have even `ArgumentOutOfRangeException` or something like this in `spectre::core::exception`. It could be of not too bad use here. (You can specify `data.size()` as bad parameter).

For the future: do not provide user-oriented messages at exception throwing site. Exception should gather the context in such a manner that does not violate DRY and is maintainable. Custom messages at each throw site aren't.